apps/api/src/yjs/v2/executor/executor.ts

```diff
@@ -0,0 +1,661 @@
+            )
+
+            const tick = async () => {
+              try {
+                if (!this.isRunning) {
+                  logger().debug(
+                    {
+                      port: process.env['PORT'],
+                      id: this.id,
+                      docId: this.docId,
+                      workspaceId: this.workspaceId,
+                      documentId: this.documentId,
+                    },
+                    'Executor is not running. Stopping consumer loop.'
+                  )
+                  resolve()
+                  return
+                }
+
+                const currentBatch = this.queue.getCurrentBatch()
+                if (!currentBatch) {
+                  this.timeout = setTimeout(() => tick(), 500)
+                  return
+                }
+
+                logger().trace(
+                  {
+                    port: process.env['PORT'],
+                    id: this.id,
+                    docId: this.docId,
+                    workspaceId: this.workspaceId,
+                    documentId: this.documentId,
+                    queueLength: this.queue.length,
+                  },
+                  'Making execution queue progress'
+                )
+
+                this.currentExecution = this.makeBatchProgress(currentBatch)
+                await this.currentExecution
+                if (this.isRunning) {
+                  this.timeout = setTimeout(() => tick(), 0)
+                }
+              } catch (err) {
+                reject(err)
+              }
+            }
+
+            tick()
+          })
+      )
+    } catch (err) {
+      logger().error(
+        {
+          port: process.env['PORT'],
+          id: this.id,
+          docId: this.docId,
+          workspaceId: this.workspaceId,
+          documentId: this.documentId,
+          err,
+        },
+        'Unexpected error while executing queue. Retrying in 2 seconds'
+      )
+      setTimeout(() => this.execute(), 2000)
+    }
+  }
+
+  private async makeBatchProgress(batch: ExecutionQueueBatch) {
+    logger().trace(
+      {
+        port: process.env['PORT'],
+        id: this.id,
+        docId: this.docId,
+        workspaceId: this.workspaceId,
+        documentId: this.documentId,
+        batchStatus: batch.status,
+        batchRemaining: batch.remaining,
+        batchLength: batch.length,
+      },
+      'Making batch progress'
+    )
+
+    const current = batch.getCurrent()
+    if (!current) {
+      logger().trace(
+        {
+          port: process.env['PORT'],
+          id: this.id,
+          docId: this.docId,
+          workspaceId: this.workspaceId,
+          documentId: this.documentId,
+        },
+        'Got to the end of current batch, advancing queue to next batch'
+      )
+      this.queue.advance()
+      return
+    }
+
+    const status = current.getStatus()
+
+    switch (status._tag) {
+      case 'running':
```

Comment:
Write a TODO here to remember that eventually we should do something better if we get a item that is already executing